Exception when trying to update an int value from the client to the model in Java

If a property is defined as Number in client template (.html) and it tries to update its conterpart field in the Java Model that is also declared as a supported type, as an int, I get this exception:

```
Caused by: java.lang.IllegalArgumentException: The stored model value '1' type 'java.lang.String' cannot be used as a type for a model property with type 'int'
	at com.vaadin.flow.templatemodel.BasicModelType.modelToApplication(BasicModelType.java:64)
```

Source code used to reproduce the error:

```
<dom-module id="x-inter">
	<template>
		<input type="radio" name="toggle" id="id1" value= 1  on-change="_getSelected">One
		<input type="radio" name="toggle" id="id2" value= 2  on-change="_getSelected">Two
	</template>
	<script>
		class XInter extends Polymer.Element{
			static get is(){
				return "x-inter";
			}
			static get properties(){
				return {
					selectedIndex: {
						type: Number,
						reflectToAttribute: true,
						notify: true
					}
				};	
			}
			_getSelected(){
				if(this.shadowRoot.querySelector('input[name=toggle]:checked')){
					this.selectedIndex = this.shadowRoot.querySelector('input[name=toggle]:checked').value;
				}
				else
					this.selectedIndex = -1;
				this.dispatchEvent(new CustomEvent('change', {}));
			}
		}
		customElements.define(XInter.is, XInter);
	</script>
</dom-module>
```

```
<dom-module id="x-element">
	<template>
		<x-inter selected-index= {{selectedValue}} on-change="_valueChanged"></x-inter>
	</template>
	<script>
		class XElement extends Polymer.Element{
			static get is(){
				return "x-element";
			}
			static get properties(){
				return {
					selectedValue:{
						type: Number,
						notify: true
					}	
				};
			}
		}
		customElements.define(XElement.is, XElement);
	</script>
</dom-module>
```

```
@Tag("x-element")
@HtmlImport("x-element.html")
public class XElement extends PolymerTemplate<Model> {

  public XElement() {
    this.getElement().synchronizeProperty("selectedValue", "_valueChanged");
    this.getElement().addPropertyChangeListener("selectedValue", e -> {
      System.out.println(getModel().getSelectedValue());
    });
  }

  public interface Model extends TemplateModel {
    int getSelectedValue();
    void setSelectedValue(int value);
  }
}
```

--- a/flow-server/src/main/java/com/vaadin/flow/templatemodel/AbstractBasicModelType.java
+++ b/flow-server/src/main/java/com/vaadin/flow/templatemodel/AbstractBasicModelType.java
@@ -74,6 +74,45 @@ public abstract class AbstractBasicModelType<T> implements ModelType {
         }
     }
 
+    /**
+     * Converts the given model value to the application type of this model
+     * type. The conversion automatically handles default values for primitive
+     * types and conversion of client-originiated numbers to the expected Java
+     * number type.
+     *
+     * @param modelValue
+     *            the model value to convert
+     * @return the converted value, not <code>null</code> if the application
+     *         type is a primitive
+     */
+    protected Object convertToApplication(Serializable modelValue) {
+        if (modelValue == null && getJavaType().isPrimitive()) {
+            return ReflectTools.getPrimitiveDefaultValue(getJavaType());
+        }
+        if (modelValue == null) {
+            return null;
+        }
+
+        Class<?> convertedJavaType = ReflectTools
+                .convertPrimitiveType(getJavaType());
+
+        // Numeric value from the client is always Double
+        if (modelValue instanceof Double
+                && convertedJavaType == Integer.class) {
+            modelValue = Integer.valueOf(((Double) modelValue).intValue());
+        }
+
+        if (convertedJavaType == modelValue.getClass()) {
+            return modelValue;
+        } else {
+            throw new IllegalArgumentException(String.format(
+                    "The stored model value '%s' type '%s' "
+                            + "cannot be used as a type for a model property with type '%s'",
+                    modelValue, modelValue.getClass().getName(),
+                    getJavaType().getName()));
+        }
+    }
+
     protected static <M> Map<Class<?>, M> loadBasicTypes(
             Function<Class<?>, M> factory) {
         Map<Class<?>, M> map = Stream

--- a/flow-server/src/main/java/com/vaadin/flow/templatemodel/BasicComplexModelType.java
+++ b/flow-server/src/main/java/com/vaadin/flow/templatemodel/BasicComplexModelType.java
@@ -21,7 +21,6 @@ import java.util.Collections;
 import java.util.Map;
 import java.util.Optional;
 
-import com.vaadin.flow.internal.ReflectTools;
 import com.vaadin.flow.internal.StateNode;
 import com.vaadin.flow.internal.nodefeature.BasicTypeValue;
 
@@ -82,24 +81,8 @@ public class BasicComplexModelType<T> extends AbstractBasicModelType<T>
         assert stateNode.hasFeature(BasicTypeValue.class);
         Serializable value = stateNode.getFeature(BasicTypeValue.class)
                 .getValue();
-        if (value == null && getJavaType().isPrimitive()) {
-            return getJavaType()
-                    .cast(ReflectTools.getPrimitiveDefaultValue(getJavaType()));
-        }
-        if (value == null) {
-            return null;
-        }
-        Class<?> boxedType = ReflectTools.convertPrimitiveType(getJavaType());
-        if (boxedType
-                .equals(ReflectTools.convertPrimitiveType(value.getClass()))) {
-            return (T) boxedType.cast(value);
-        } else {
-            throw new IllegalArgumentException(String.format(
-                    "The stored model value '%s' type '%s' "
-                            + "cannot be used as a type for a model property with type '%s'",
-                    value, value.getClass().getName(),
-                    getJavaType().getName()));
-        }
+
+        return (T) convertToApplication(value);
     }
 
     @Override

--- a/flow-server/src/main/java/com/vaadin/flow/templatemodel/BasicModelType.java
+++ b/flow-server/src/main/java/com/vaadin/flow/templatemodel/BasicModelType.java
@@ -19,8 +19,6 @@ import java.io.Serializable;
 import java.util.Map;
 import java.util.Optional;
 
-import com.vaadin.flow.internal.ReflectTools;
-
 /**
  * A model type representing an immutable leaf value, e.g. strings, numbers or
  * booleans.
@@ -50,22 +48,7 @@ public class BasicModelType extends AbstractBasicModelType {
 
     @Override
     public Object modelToApplication(Serializable modelValue) {
-        if (modelValue == null && getJavaType().isPrimitive()) {
-            return ReflectTools.getPrimitiveDefaultValue(getJavaType());
-        }
-        if (modelValue == null) {
-            return null;
-        }
-        if (ReflectTools.convertPrimitiveType(getJavaType()).equals(
-                ReflectTools.convertPrimitiveType(modelValue.getClass()))) {
-            return modelValue;
-        } else {
-            throw new IllegalArgumentException(String.format(
-                    "The stored model value '%s' type '%s' "
-                            + "cannot be used as a type for a model property with type '%s'",
-                    modelValue, modelValue.getClass().getName(),
-                    getJavaType().getName()));
-        }
+        return convertToApplication(modelValue);
     }
 
     @Override

--- a/flow-server/src/test/java/com/vaadin/flow/templatemodel/BeanModelTypeTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/templatemodel/BeanModelTypeTest.java
@@ -24,8 +24,11 @@ import java.util.stream.Collectors;
 import org.junit.Assert;
 import org.junit.Test;
 
+import com.vaadin.flow.internal.JsonCodec;
 import com.vaadin.flow.internal.StateNode;
 import com.vaadin.flow.internal.nodefeature.ElementPropertyMap;
+
+import elemental.json.Json;
 
 public class BeanModelTypeTest {
     // Partial overlap with Bean
@@ -279,6 +282,38 @@ public class BeanModelTypeTest {
         Assert.assertEquals("3", bean.getString());
         Assert.assertEquals(3, bean.getIntValue());
         Assert.assertNull(bean.getIntObject());
+    }
+
+    @Test
+    public void clientValueToApplication() {
+        BeanModelType<Bean> beanType = new BeanModelType<>(Bean.class,
+                PropertyFilter.ACCEPT_ALL, false);
+
+        ElementPropertyMap model = createEmptyModel();
+        model.setProperty("doubleValue",
+                JsonCodec.decodeWithoutTypeInfo(Json.create(3)));
+        model.setProperty("doubleObject",
+                JsonCodec.decodeWithoutTypeInfo(Json.create(3)));
+        model.setProperty("intValue",
+                JsonCodec.decodeWithoutTypeInfo(Json.create(3)));
+        model.setProperty("intObject",
+                JsonCodec.decodeWithoutTypeInfo(Json.create(3)));
+        model.setProperty("booleanValue",
+                JsonCodec.decodeWithoutTypeInfo(Json.create(true)));
+        model.setProperty("booleanObject",
+                JsonCodec.decodeWithoutTypeInfo(Json.create(true)));
+        model.setProperty("string",
+                JsonCodec.decodeWithoutTypeInfo(Json.create("3")));
+
+        Bean bean = beanType.modelToApplication(model.getNode());
+
+        Assert.assertEquals(3.0, bean.getDoubleValue(), 0);
+        Assert.assertEquals(3.0, bean.getDoubleObject(), 0);
+        Assert.assertEquals(3, bean.getIntValue());
+        Assert.assertEquals(3, bean.getIntObject().intValue());
+        Assert.assertEquals(Boolean.TRUE, bean.getBooleanObject());
+        Assert.assertTrue(bean.isBooleanValue());
+        Assert.assertEquals("3", bean.getString());
     }
 
     @Test

--- a/flow-server/src/test/java/com/vaadin/flow/templatemodel/ModelDescriptorTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/templatemodel/ModelDescriptorTest.java
@@ -366,6 +366,17 @@ public class ModelDescriptorTest {
     }
 
     @Test
+    public void basicComplexModelType_intAndInteger_numberFromClient() {
+        ComplexModelType<?> type = BasicComplexModelType.get(Integer.class)
+                .get();
+
+        StateNode stateNode = type.applicationToModel(2.0d, null);
+        Object applicationValue = type.modelToApplication(stateNode);
+
+        Assert.assertEquals(Integer.valueOf(2), applicationValue);
+    }
+
+    @Test
     public void basicComplexModelType_booleabAndBoolean() {
         assertComplexModeType(boolean.class, true, false);
         assertComplexModeType(Boolean.class, false, true);

